### ISSUE TITLE
WPF and WinForms templates should fix project files

### ISSUE THREAD

**Issue by davidwengier** [dotnet/project-system#4718]
When adding a WPF or WinForms item template to an SDK-style project, the template should:

* Update the project file SDK to be Microsoft.NET.Sdk.WindowsDesktop if not present
* Add <UseWPF>true</UseWPF> or <UseWinforms>true</UseWinforms> if not present

**Comment by jmarolf** [comment]
the naive way I am imagining implementing this is:

1. When an item template attempts to add assemblies to the project, have the project system intercept this and examine the reference (@davkean 	or @lifengl is there an extenion point for this today?)
2. If the assembly would come from framework references that require Windows Desktop support, modify the project file to use `Microsoft.NET.Sdk.WindowsDesktop`
3. Add `UseWPF` or `UseWinForms` depending on the reference (@davidwengier 	is there a way to easily determine this from capabilites?)

**Comment by davkean** [comment]
This is blocked on me. I don't want any of this business logic built into the project-system at all, project-systems span SDKs/platforms, as this logic changes (and it will), we'll need to handle all the different incarnations that the SDL supports. If we cannot improve the project format to make it entirely FrameworkReference based in 16.3, I would cut this feature.

**Comment by jmarolf** [comment]
Thanks @davkean 	I'll assign this over to you then.  Let me know if I can help.

**Comment by nguerrera** [comment]
Probably not what was implied above, but the way I see it, there are options other than changing the project format or hard-coding in the SDK. For example, I could imagine item templates carrying information about required properties, required sdks...

**Comment by nguerrera** [comment]
Even if it were entirely FrameworkReference-based, you'd need the template to be able to signal what FrameworkReference needs to be added, right? Why is that workable, but not SDK or property change?

**Comment by davkean** [comment]
Not following?

**Comment by nguerrera** [comment]
You wrote:

> If we cannot improve the project format to make it entirely FrameworkReference based

I don't understand why a template that can say "add a framework reference to X" is all that different than a template that can say "set SDK to Y and property Z to W".  I don't get the leap to FrameworkReference-is-the-only-thing. I'm just saying that it's conceivable to neither hard code nor use FrameworkReference as the key to all build logic required.

**Comment by nguerrera** [comment]
Is it just about timelines difficulty or am I missing a fundamental problem?

**Comment by davkean** [comment]
Mapping a template -> a reference or SDK of some kind is fundamentally easy, and something we've always had. Template has ability to specify a reference, this gets translated to effectively References.Add on the project, which directly translates to adding an item in the project. It doesn't differ based what state the project is in or whether it has different imports, etc. It just adds the reference.

To handle above and the other situations [called out here](https://github.com/dotnet/project-system/issues/4873), "something" has to know the rules about when to add/unset a property VS & SDK , versus when to just add (or remove) a framework reference (which differs based on what SDK you have installed), and this will change over time and SDK versions as we bring up UWP, Xamarin and other technologies. That logic does not below in a platform-agnostic project system, it should ship in the SDK where the business rules themselves live.

**Comment by nguerrera** [comment]
Thanks.. I am 100% in agreement with "That logic does not below in a platform-agnostic project system", btw. I was never suggesting to hard code anything in project system, just trying to understand more about how it might be possible to encode the current set of changes required in the template. The point about the statefulness of it vs. always just adding a set of references clarifies the challenge for me.

**Comment by davkean** [comment]
It's more than just templates too as https://github.com/dotnet/project-system/issues/4873 calls out - we have a literal project-agnostic and CPS-based APIs that enables enumerating, adding and removing references (assembly references, project references, package references, SDK references, COM references) that have a map between "add this reference" -> "add this item", "remove this reference" -> "remove this item". FrameworkReference does not fit into that mold, and I'm hesitant in adding these APIs as I feel that this business logic will change as we bring up more the other platforms.

**Comment by jmarolf** [comment]
@nguerrera 	Its mostly about timelines / difficulty from my side.  I'd need to design and push a new shell API from the VS side so that templates could declare framework references (I still don't know how I build the mapping from framework-reference-name -> set-of-changes-that-the-project-file-needs but lets side-step that one for a bit)

For nuget we punted on this way back and had a template wizard that template authors could use (shipped with VS) that allowed them to add nuget packages in their templates.  We could use the same approach for framework references, but my concern is that means that all template authors will have item templates that don't work on Core by default.  Today, if you have `UseWPF` set to true and you try and add a template that needs WPF references, the project system sees that the references are already there and assumes it has nothing to do so things just work.  Framework references, at least as regards templates, ask that all template authors update things specifically so their item templates work on Core.

**Comment by nguerrera** [comment]
All good points. 

On the subject of other stacks coming online, correct me if I am wrong, but I think we've never made it possible before to add uwp or xanarin templates to a vanilla .net project. Ditto web, right? It is really only wpf and winforms that were first class in this sense, right?

**Comment by davkean** [comment]
Templates, yes[1].

They were new project types, but for ".NET 5" we're coming back to a single stack. Ideally, you'll want to start with a class library and add templates from any tech that supports ".NET 5" rather than back to a different library template for each technology.

[1] Xamarin Forms and XAML islands makes this a little unclear. The other thing to remember is that WPF templates don't currently show up in "Windows Forms" projects (with exception to 1 which is a bug).

**Comment by jmarolf** [comment]
> but I think we've never made it possible before to add uwp or xanarin templates to a vanilla .net project.

Correct, but I agree with @davkean 	that in the .NET 5 time-frame I would like to have a story that allows templates to "just work" for scenarios where I am not running a custom wizard that is doing who-know-what. Today I don't know how to go about doing this other than 1.) making some "tactical fixes" in the project system to help translate between the templates intent and the sdk or 2.) creating an entirely new API that allows a template author to express this intent, but now people need to fork their item templates for .NET Core.
